preserving data for Repeating groups.

rahimkhalid opened the pull request:
[Fixes #333 ](https://github.com/OpenSRP/opensrp-client-native-form/issues/333)
**-** Made repeating group values preserved so that once data is entered, it is preserved and selected value can be returned to the user If any validation has to be made.
**-** Fixed placement of repeated group section.

rahimkhalid commented:
@githengi

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/task/AttachRepeatingGroupTask.java:
```diff
@@ -150,7 +155,24 @@ private LinearLayout buildRepeatingGroupLayout(ViewParent parent) throws Excepti
+                    JSONObject fieldObject = fields.getJSONObject(fieldIndex);
+                    updatedFields.put(fieldObject);
+                }
+                step.put(FIELDS, updatedFields);
```
It's not very clear what is happening here, could you please add comments that explain the algorithm?

rahimkhalid replied:
I have added in-line comments explaining the logic.

vincent-karuri replied:
Cool, more understandable now, thanks.

I think we should store the `repeatingGroupIndex` as a global variable, initially equal to `-1` then after it's initialized to it's true value, we won't need to do this traversal again (which is expensive).

So we can factor out this logic to a separate function and only call the function if `repeatingGroupIndex == 1`?

vincent-karuri replied:
Re-initialization to `-1` would happen in the `getViewsFromJson` function.

vincent-karuri replied:
Could the logic move to a separate private function with a descriptive name for what is happening?

The helpful docstring you added can then be put above the function signature.

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java:
```diff
@@ -66,6 +67,14 @@
+
+        }
+
+        api = (JsonApi) formFragment.getContext();
```
The `JsonApi` can be retrieved from the `WidgetArgs` : `(JsonApi) widgetArgs.getContext()`

rahimkhalid replied:
Fixed.

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java:
```diff
@@ -142,6 +152,17 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
+                if(repeatingObject.has(JsonFormConstants.TEXT)) {
+                    try {
+                        repeatingObject.put(JsonFormConstants.TEXT,s.toString());
+                        widgetArgs.setJsonObject(repeatingObject);
```
This line is not necessary.

It can be `widgetArgs.getJsonObject().put(JsonFormConstants.TEXT,s.toString());`

rahimkhalid replied:
Fixed.

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java:
```diff
@@ -142,6 +152,17 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
                 doneButton.setImageResource(R.drawable.ic_done_grey);
                 JsonFormFragmentPresenter.validate(widgetArgs.getFormFragment(), referenceEditText, false);
+                JSONObject repeatingObject = widgetArgs.getJsonObject();
+                if(repeatingObject.has(JsonFormConstants.TEXT)) {
```
Why is this check required?

rahimkhalid replied:
For repeating group, If you need to know selected repeating group count value, add text field in JSON form. It is necessary to check if repeating the group contains the TEXT field so that it shouldn't update the values field. According to write Values working.
e.g
`{
        "key": "dips",
        "type": "repeating_group",
        "text": "", ...`
The text value will gets updated with user selected count. So if s/he needs to make any decision based on that.

vincent-karuri replied:
Also, the approach used in most of the library is :

```
String repeatingGroupCount = repeatingObject.optString(JsonFormConstants.TEXT);
if (repeatingGroupCount  != null) {
       ...
}
```

instead of using `repeatingObject.has(JsonFormConstants.TEXT)` and `repeatingObject.get(JsonFormConstants.TEXT)`. Better to use the same for consistency.

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/interactors/JsonFormInteractor.java:
```diff
@@ -173,16 +181,42 @@ public void fetchFields(List<View> viewsFromJson, String stepName, JsonFormFragm
+                            break;
+                        }
+                    }
+                }
```
Could you please give more information on what is being done here?

rahimkhalid replied:
Adding Repeating group child section just after repeating group So that when you come back to the repeated group section step, all children should be visible after that Repeating group.
FIXED: Should add repeating group sections after repeating the group field, not at the end step.

rahimkhalid commented:
@vincent-karuri Sorry for delayed reply but thanks for pointing out concerned code. I have fixed all the discussions and added comments on some complex logic.

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/event/ExpansionPanelInflateEvent.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ExpansionPanelInflateEvent extends BaseEvent {
```
This class looks almost identical to `RefreshExpansionPanelEvent`. 

Wouldn't a better approach be to factor out the similar components to the `BaseEvent`?

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java:
```diff
@@ -142,6 +150,34 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
+                if(repeatingObject.has(JsonFormConstants.TEXT)) {
+                    try {
+                        repeatingObject.put(JsonFormConstants.TEXT,s.toString());
+                        ((JsonApi) widgetArgs.getContext()).writeValue(widgetArgs.getStepName(), widgetArgs.getJsonObject().getString(KEY),s.toString(),"", "", "", widgetArgs.isPopup());
```
Doesn't this overwrite the value of the repeating group json?

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/task/AttachRepeatingGroupTask.java:
```diff
@@ -150,7 +154,32 @@ private LinearLayout buildRepeatingGroupLayout(ViewParent parent) throws Excepti
+                for(int fieldIndex = 0; fieldIndex < fields.length(); fieldIndex++) {
+                    JSONObject fieldObject = fields.getJSONObject(fieldIndex);
+                    updatedFields.put(fieldObject);
+                    if(fieldObject.has(KEY) && fieldObject.getString(KEY).equals(repeatingGroupKey)) {
```
Prefer assigning `fieldObject.optString(KEY) ` to a variable and check if its `null`  and equal to `repeatingGroupKey`.

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/interactors/JsonFormInteractor.java:
```diff
@@ -173,16 +180,42 @@ public void fetchFields(List<View> viewsFromJson, String stepName, JsonFormFragm
+                            break;
+                        }
+                    }
+                }
```
This may also need commenting and maybe moving it to a separate function, with a descriptive name for what is happening.

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/interactors/JsonFormInteractor.java:
```diff
@@ -173,16 +180,42 @@ public void fetchFields(List<View> viewsFromJson, String stepName, JsonFormFragm
+                if (childJson.get(TYPE).equals(REPEATING_GROUP)) {
+                    for (int j = i + 1; j < fields.length(); j++) {
+                        JSONObject repeatingChildJson = fields.getJSONObject(j);
+                        if (repeatingChildJson.has(PARENT_REPEATING_GROUP) && repeatingChildJson.getString(PARENT_REPEATING_GROUP).equals(childJson.getString(KEY))) {
```
Prefer assigning `repeatingChildJson.optString(PARENT_REPEATING_GROUP)` to a variable and check if its null and equal to `childJson.getString(KEY`.

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/interactors/JsonFormInteractor.java:
```diff
@@ -173,16 +180,42 @@ public void fetchFields(List<View> viewsFromJson, String stepName, JsonFormFragm
 
             for (int i = 0; i < fields.length(); i++) {
                 JSONObject childJson = fields.getJSONObject(i);
```
Could this be renamed to `fieldJson` for clarity?

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/interactors/JsonFormInteractor.java:
```diff
@@ -173,16 +180,42 @@ public void fetchFields(List<View> viewsFromJson, String stepName, JsonFormFragm
                         listener, popup);
+                if (childJson.get(TYPE).equals(REPEATING_GROUP)) {
+                    for (int j = i + 1; j < fields.length(); j++) {
+                        JSONObject repeatingChildJson = fields.getJSONObject(j);
```
Maybe rename to `repeatingGroupFieldJson` for clarity?

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/interactors/JsonFormInteractor.java:
```diff
@@ -104,7 +111,7 @@ protected void registerWidgets() {
         map.put(JsonFormConstants.SPACER, new ComponentSpacerFactory());
         map.put(JsonFormConstants.NATIVE_EDIT_TEXT, new NativeEditTextFactory());
         map.put(JsonFormConstants.TIME_PICKER, new TimePickerFactory());
-        map.put(JsonFormConstants.REPEATING_GROUP, new RepeatingGroupFactory());
```
Avoid re-formatting these since it's inconsistent with the formatting in the rest of the class.

vincent-karuri commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/interactors/JsonFormInteractor.java:
```diff
@@ -173,16 +180,42 @@ public void fetchFields(List<View> viewsFromJson, String stepName, JsonFormFragm
+                List<View> views = fetchViews(viewsFromJson, stepName, formFragment, childJson.getString(TYPE), childJson,
                         listener, popup);
+                if (childJson.get(TYPE).equals(REPEATING_GROUP)) {
+                    for (int j = i + 1; j < fields.length(); j++) {
```
Why not use a `while` loop here instead of having a nested `if` statement with an `else` that does not contain logic?

vincent-karuri commented:
Also, please address general formatting issues like following Android Studio conventions for spacing.

I found [this](https://stackoverflow.com/a/16580200/9782187) helpful.

dubdabasoduba commented:
@rahimkhalid is this still required?

rahimkhalid commented:
@dubdabasoduba no it's not required. So closing this PR.

dubdabasoduba commented:
👍
